docs/developer/utils.rst

```diff
@@ -0,0 +1,495 @@
+  the first time. It shall be set to false afterwards.
+- ``tolerance_crossed``: it will be set to true if the metric has crossed
+  the threshold for tolerance configured in alert settings. Otherwise, it
+  will be set to false.
+
+``first_time`` parameter can be used to avoid initiating unneeded actions.
+For example, sending recovery notifications.
+
+This signal is emitted when the threshold value of a ``Metric`` defined in
+alert settings is crossed.
+
+``pre_metric_write``
+~~~~~~~~~~~~~~~~~~~~
+
+**Path**: ``openwisp_monitoring.monitoring.signals.pre_metric_write``
+
+**Arguments**:
+
+- ``metric``: ``Metric`` object whose data shall be stored in timeseries
+  database
+- ``values``: metric data that shall be stored in the timeseries database
+- ``time``: time with which metrics will be saved
+- ``current``: whether the data has just been collected or was collected
+  previously and sent now due to network connectivity issues
+
+This signal is emitted for every metric before the write operation is sent
+to the timeseries database.
+
+``post_metric_write``
+~~~~~~~~~~~~~~~~~~~~~
+
+**Path**: ``openwisp_monitoring.monitoring.signals.post_metric_write``
+
+**Arguments**:
+
+- ``metric``: ``Metric`` object whose data is being stored in timeseries
+  database
+- ``values``: metric data that is being stored in the timeseries database
+- ``time``: time with which metrics will be saved
+- ``current``: whether the data has just been collected or was collected
+  previously and sent now due to network connectivity issues
+
+This signal is emitted for every metric after the write operation is
+successfully executed in the background.
+
+Management commands
+-------------------
+
+.. _run_checks:
+
+``run_checks``
+~~~~~~~~~~~~~~
+
+This command will execute all the `available checks <checks>`_ for all the
+devices. By default checks are run periodically by *celery beat*.
+
+Example usage:
+
+.. code-block:: shell
+
+    cd tests/
+    ./manage.py run_checks
+
+``migrate_timeseries``
+~~~~~~~~~~~~~~~~~~~~~~
+
+This command triggers asynchronous migration of the time-series database.
+
+Example usage:
+
+.. code-block:: shell
+
+    cd tests/
+    ./manage.py migrate_timeseries
+
+Exceptions
+----------
+
+``TimeseriesWriteException``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Path**: ``openwisp_monitoring.db.exceptions.TimeseriesWriteException``
+
+If there is any failure due while writing data in timeseries database,
+this exception shall be raised with a helpful error message explaining the
+cause of the failure. This exception will normally be caught and the
+failed write task will be retried in the background so that there is no
+loss of data if failures occur due to overload of Timeseries server. You
+can read more about this retry mechanism at
+`OPENWISP_MONITORING_WRITE_RETRY_OPTIONS
+<#openwisp-monitoring-write-retry-options>`_.
+
+``InvalidMetricConfigException``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Path**:
```

Comment:
```suggestion
**Full python path**:
```
I've used the same in OpenWISP Users. Applies to all occurrences (use search and replace).